File: exercises/csharp/url-monitor-cli_code_review/src/UrlMonitor/Notifier.cs
namespace UrlMonitor;

public static class Notifier
{
    /// <summary>
    /// Return the current UTC time in ISO 8601 format.
    /// </summary>
    public static string Iso8601Now() =>
        DateTime.UtcNow.ToString("yyyy-MM-ddTHH:mm:ssZ");

    /// <summary>
    /// Log an informational message to stdout.
    /// </summary>
    public static void LogInfo(string message)
    {
        Console.WriteLine($"{Iso8601Now()} INFO  {message}");
    }

    /// <summary>
    /// Log an error message to stderr.
    /// </summary>
    public static void LogError(string message)
    {
        Console.Error.WriteLine($"{Iso8601Now()} ERROR {message}");
    }

    /// <summary>
    /// Emit a status transition notification.
    /// </summary>
    public static void EmitTransition(string url, Status newStatus, CheckResult result)
    {
        var statusLabel = newStatus == Status.Up ? "UP   " : "DOWN ";
        var detail = FormatDetail(result);
        Console.WriteLine($"{Iso8601Now()} {statusLabel} {url}  ({detail})");
    }

    /// <summary>
    /// Emit a verbose check log line.
    /// </summary>
    public static void EmitVerboseCheck(string url, Status status, CheckResult result)
    {
        var statusLabel = status == Status.Up ? "UP   " : "DOWN ";
        var detail = FormatDetail(result);
        Console.WriteLine($"{Iso8601Now()} {statusLabel} {url}  ({detail})");
    }

    private static string FormatDetail(CheckResult result)
    {
        if (result.Error is not null)
        {
            return $"error: {result.Error}";
        }
        return $"HTTP {result.HttpStatus}, {result.TotalMs:F0}ms";
    }
}

File: solutions/python/url-monitor-cli/src/url_monitor/notifier.py
"""Stdout logging, transition emission, and timestamp utilities."""

from __future__ import annotations

import sys
from datetime import datetime, timezone

from url_monitor.checker import CheckResult, Status, status_name


def iso8601_now() -> str:
    """Return current UTC time in ISO 8601 format."""
    return datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")


def log_info(msg: str) -> None:
    """Log an informational message to stdout."""
    ts = iso8601_now()
    print(f"{ts} {msg}", flush=True)


def log_error(msg: str) -> None:
    """Log an error message to stderr."""
    ts = iso8601_now()
    print(f"{ts} ERROR {msg}", file=sys.stderr, flush=True)


def _result_detail(result: CheckResult) -> str:
    """Format check result detail string."""
    if result.error is not None:
        return f"(error: {result.error})"
    return f"(HTTP {result.http_status}, {result.total_ms:.0f}ms)"


def emit_transition(
    url: str, prev: Status, now: Status, result: CheckResult
) -> None:
    """Emit a status transition notification."""
    status_label = status_name(now).upper()
    detail = _result_detail(result)
    ts = iso8601_now()
    line = f"{ts} {status_label:<5} {url}  {detail}"
    print(line, flush=True)


def log_check(url: str, status: Status, result: CheckResult) -> None:
    """Log a check result (verbose mode)."""
    status_label = status_name(status).upper()
    detail = _result_detail(result)
    log_info(f"{status_label:<5} {url}  {detail}")
